applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,878 @@
+// direction is predetermined by the size of the conditions constituting the
+// trailing edge.
+void Define3DWakeProcess::ShedWakeSurfaceFromTheTrailingEdge()
+{
+    const Properties::Pointer pElemProp = mrStlWakeModelPart.pGetProperties(0);
+    const double number_of_elements = mSheddedWakeDistance / mSheddedWakeElementSize;
+    const unsigned int number_of_elements_in_wake_direction = int(number_of_elements);
+
+    IndexType node_index = 0;
+    IndexType element_index = 0;
+
+    array_1d<double,3> coordinates1 = ZeroVector(3);
+    array_1d<double,3> coordinates2 = ZeroVector(3);
+    array_1d<double,3> coordinates3 = ZeroVector(3);
+    array_1d<double,3> coordinates4 = ZeroVector(3);
+
+    for (auto& r_cond : mrTrailingEdgeModelPart.Conditions()) {
+        auto& r_geometry = r_cond.GetGeometry();
+        coordinates1 = r_geometry[0].Coordinates();
+        coordinates2 = r_geometry[1].Coordinates();
+
+        if(r_geometry[0].GetValue(WING_TIP) != 0){
+            DecreaseWakeWidthAtTheWingTips(coordinates1, coordinates2);
+        }
+        else if(r_geometry[1].GetValue(WING_TIP) != 0){
+            DecreaseWakeWidthAtTheWingTips(coordinates2, coordinates1);
+        }
+
+        coordinates3 = coordinates1 + mSheddedWakeElementSize * mWakeDirection;
+        coordinates4 = coordinates2 + mSheddedWakeElementSize * mWakeDirection;
+
+
+        CreateWakeSurfaceNodesAndElements(node_index, coordinates1, coordinates2, coordinates3,
+                                   coordinates4, element_index, pElemProp);
+
+        for (unsigned int j = 0; j < number_of_elements_in_wake_direction; j++){
+            coordinates1 = coordinates3;
+            coordinates2 = coordinates4;
+            coordinates3 = coordinates1 + mSheddedWakeElementSize * mWakeDirection;
+            coordinates4 = coordinates2 + mSheddedWakeElementSize * mWakeDirection;
+
+            CreateWakeSurfaceNodesAndElements(node_index, coordinates1, coordinates2, coordinates3,
+                                   coordinates4, element_index, pElemProp);
+        }
+    }
+}
+
+void Define3DWakeProcess::DecreaseWakeWidthAtTheWingTips(array_1d<double, 3>& rPoint1,
+                                                         const array_1d<double, 3>& rPoint2)
+{
+    const auto& te_edge = rPoint1 - rPoint2;
+    const double projection = inner_prod(te_edge, mSpanDirection);
+
+    if (projection > 0.0)
+    {
+        rPoint1 -= 1e-6 * te_edge;
+    }
+    else
+    {
+        rPoint1 += 1e-6 * te_edge;
+    }
+}
+
+void Define3DWakeProcess::CreateWakeSurfaceNodesAndElements(
+    IndexType& rNode_index,
+    const array_1d<double, 3>& rCoordinates1,
+    const array_1d<double, 3>& rCoordinates2,
+    const array_1d<double, 3>& rCoordinates3,
+    const array_1d<double, 3>& rCoordinates4,
+    IndexType& rElement_index,
+    const Properties::Pointer pElemProp) const
+{
+    const std::vector<ModelPart::IndexType> nodes_ids = CreateWakeSurfaceNodes(
```

Comment:
```suggestion
    const std::array<ModelPart::IndexType, 4> nodes_ids = CreateWakeSurfaceNodes(
```
We know this will be always length 4 no?

Reply:
True!